### PR TITLE
Add MeshBase::active_local_element_ptr_range() 

### PR DESCRIPTION
**Pull request by jwpeterson**
This removes the majority of the remaining mesh iterator boilerplate code. 

I count about 150 other miscellaneous mesh iterator loops (`flagged_pid_elements_begin()`, etc.) but I don't plan on fixing them since the net number of lines saved won't be as large.

### CODE REVIEW

File: src/error_estimation/adjoint_refinement_estimator.C [hunk 1]
```diff
@@ -370,93 +370,72 @@ void AdjointRefinementEstimator::estimate_error (const System & _system,
...
-  // this processor
-  {
-    MeshBase::const_element_iterator elem_it = mesh.active_local_elements_begin();
-    const MeshBase::const_element_iterator elem_end = mesh.active_local_elements_end();
-
-    // Start loop over elems
-    for (; elem_it != elem_end; ++elem_it)
+  // this processor.
+  for (const auto & elem : mesh.active_local_element_ptr_range())
+    for (unsigned int n=0; n != elem->n_nodes(); ++n)
```
Comment (roystgnr) [inline src/error_estimation/adjoint_refinement_estimator.C]: Huh.  Why didn't I change this to a node_ref_range()?

File: src/error_estimation/adjoint_refinement_estimator.C [hunk 2]
```diff
@@ -529,18 +502,12 @@ void AdjointRefinementEstimator::estimate_error (const System & _system,
...
 
   for (std::size_t j=0; j != system.qoi.size(); j++)
-    {
-      if (_qoi_set.has_index(j))
-        {
-          delete coarse_adjoints[j];
-        }
-    }
+    if (_qoi_set.has_index(j))
+      delete coarse_adjoints[j];
```
Comment (roystgnr) [inline src/error_estimation/adjoint_refinement_estimator.C]: I'm still thrilled by the number of lines of code this PR strips, but slipping in unrelated whitespace and brace removal doesn't count for your final score.

File: src/mesh/ensight_io.C [hunk 3]
```diff
@@ -198,19 +198,13 @@ void EnsightIO::write_geometry_ascii()
...
-        ensight_parts_map[elem->type()].push_back(elem);
+  for (const auto & elem : the_mesh.active_local_element_ptr_range())
+    {
+      ensight_parts_map[elem->type()].push_back(elem);
 
-        for (unsigned int i = 0; i < elem->n_nodes(); i++)
-          mesh_nodes_map[elem->node_id(i)] = elem->point(i);
-      }
-  }
+      for (unsigned int i = 0; i < elem->n_nodes(); i++)
```
Comment (roystgnr) [inline src/mesh/ensight_io.C]: And this I should have turned into a node_index_range() already.  I'll make another pass at those later.

File: src/error_estimation/adjoint_refinement_estimator.C (hunk 1 above)
Comment (jwpeterson) [reply src/error_estimation/adjoint_refinement_estimator.C]: Maybe this one loop didn't match your regex...

File: src/error_estimation/adjoint_refinement_estimator.C (hunk 2 above)
Comment (jwpeterson) [reply src/error_estimation/adjoint_refinement_estimator.C]: The brace removal counts! Since we don't have to have the extra line for dereferencing an iterator, it allowed several extra braces to be collapsed.

File: src/mesh/ensight_io.C (hunk 3 above)
Comment (jwpeterson) [reply src/mesh/ensight_io.C]: 👍

**Comment by roystgnr** [comment]
Awesome, thanks!

I've noticed that with mesh iterators we never seemed to make the mistake of calling foo_end() inside the loop test, though...  So if we want to improve our iteration performance then we *actually* need to improve our iterator performance, there was no simple-to-fix inefficient code the way there was for the element node and element children ranges.

File: src/error_estimation/adjoint_refinement_estimator.C (hunk 2 above)
Comment (roystgnr) [reply src/error_estimation/adjoint_refinement_estimator.C]: Most of the brace removal counts, but this set was an exception.
